File: typescript/src/main/vsdk/toolkit/common/linealAlgebra/Vector2Dd.ts
export class Vector2Dd {
  public u: number;
  public v: number;

  public constructor() {
    this.u = 0.0;
    this.v = 0.0;
  }

  public static set(vector: Vector2Dd, a: number, b: number): void {
    vector.u = a;
    vector.v = b;
  }

  public static subtract(p: Vector2Dd, q: Vector2Dd, r: Vector2Dd): void {
    r.u = p.u - q.u;
    r.v = p.v - q.v;
  }

  public static add(p: Vector2Dd, q: Vector2Dd, r: Vector2Dd): void {
    r.u = p.u + q.u;
    r.v = p.v + q.v;
  }

  public static negate(p: Vector2Dd): void {
    p.u = -p.u;
    p.v = -p.v;
  }

  public static determinant(a: Vector2Dd, b: Vector2Dd): number {
    return a.u * b.v - a.v * b.u;
  }
}

File: java/base/src/main/java/vsdk/toolkit/common/linealAlgebra/Vector2Dd.java
package vsdk.toolkit.common.linealAlgebra;

public class Vector2Dd {
    public double u;
    public double v;

    public Vector2Dd() {
        u = 0.0;
        v = 0.0;
    }

    public static void set(Vector2Dd vector, double a, double b) {
        vector.u = a;
        vector.v = b;
    }

    public static void subtract(Vector2Dd p, Vector2Dd q, Vector2Dd r) {
        r.u = p.u - q.u;
        r.v = p.v - q.v;
    }

    public static void add(Vector2Dd p, Vector2Dd q, Vector2Dd r) {
        r.u = p.u + q.u;
        r.v = p.v + q.v;
    }

    public static void negate(Vector2Dd p) {
        p.u = -p.u;
        p.v = -p.v;
    }

    public static double determinant(Vector2Dd a, Vector2Dd b) {
        return a.u * b.v - a.v * b.u;
    }
}
